x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/search_knn_query_vector_builder.yml

```diff
@@ -0,0 +1,120 @@
+  - do:
+      headers:
+        Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
+        Content-Type: application/json
+      indices.forcemerge:
+        index: index-with-embedded-text
+        max_num_segments: 1
+
+  - do:
+      headers:
+        Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
+        Content-Type: application/json
+      ml.start_trained_model_deployment:
+        model_id: text_embedding_model
+        wait_for: started
+
+---
+"Test vector search with query_vector_builder":
+  - do:
+      search:
+        index: index-with-embedded-text
+        body:
+          query:
+            knn:
+              field: embedding
+              num_candidates: 3
+              query_vector_builder:
+                text_embedding:
+                  model_id: text_embedding_model
+                  model_text: "the octopus comforter smells"
```

Comment:
I realize that this is how the top level `knn` is done, but now that `inference_id` is becoming a thing, and being used in the core server code, maybe we can switch it it?

@davidkyle what do you think? Can we make an inference call directly from server now? so our query DSL turns into 

```
"knn": {
  "inference_id": "foo", 
  "my_dense_vector_field": "what did the fox jump over?"
}
```
?

Reply:
the reasoning for the `query_vector_builder` stuff (`QueryVectorBuilder` interface, same as top level `knn`) is that we cannot introduce a direct dependency between `KnnVectorQueryBuilder` (from `server`) and `TextEmbeddingQueryVectorBuilder` (from `xpack/plugin/ml`) or `InferenceAction` (from `xpach/plugin/core`).
Also it seems desirable to have consistent behaviors between top level and DSL `knn` queries.

I now see we have `SemanticTextModelSettings` in server, so if we want to "switch" to `inference_id` to be picked up by some `ml` code at runtime (which is anyway the case with `query_vector_builder`) we can probably do that, but I'd consider doing that for both top level and DSL queries eventually.

Reply:
In the near future we will want to rename `model_id` -> `inference_id` and `model_text` to ?`inference_text` but there are many places it needs to be done and should change all uses at the same time. I expect to keep the `model_id` option for compatibility.

`server` now contains the `InferenceServiceRegistry` which you can use to look up inference objects configured in _inference and call infer on those objects directly, but it doesn't have access to the models in `_ml/trained_models`. I'm assuming the purpose of this PR is to create parity between the top level knn and the knn query, in which case it is best to use the same implementation for now. 

In future we will automatically create _inference configurations for text embedding models uploaded with Eland, at that point using the `InferenceServiceRegistry` becomes an option